kmod/src/dir.c

```diff
@@ -870,6 +870,12 @@ static int scoutfs_link(struct dentry *old_dentry,
 	inode->i_ctime = dir->i_mtime;
 	inc_nlink(inode);
 
+	/* nlink 0 -> 1 (tmpfile?) but now should no longer be */
+	if (inode->i_nlink == 1) {
+		ret = remove_orphan_item(sb, scoutfs_ino(inode));
+		WARN_ON_ONCE(ret);
```

Comment:
We can handle the errors if we do this higher up.  Like we have creation and deletion wrappers in inode.c for orphan handling, we can add a item dirtying wrapper.  It'd call _item_dirty() instead of create/delete.  If you called that dirty wrapper up after dirty_inode_item then you know you have a dirty orphan in the item cache.  You could call orphan deletion after add_entry_items and know that it can't fail 'cause the item was dirty.  I'd make dirtying the orphan item and removing it conditional on some bool like "del_orphan = inode->nlink == 0" , maybe before setting up that dir_size.

Reply:
ok took a pass at this.